### PR TITLE
Add explicit output parameters to spiking output ports

### PR DESCRIPTION
**Pull request by clinssen**
Part of #946.

Add explicit output parameters to spiking output ports: if ``emit_spike(foo, bar)`` is called, then the spiking output port should be parameterised by two parameters with types equal to (or castable to) those of foo and bar.

Todo:
- [x] update documentation
- [x]  add coco to check that continuous-type output ports do not support attributes
- [ ]  <S>(?) disallow spike attribute names to alias variable names in the model scope (such as in the state or parameters block), e.g. "spike_port.foo" is not allowed when a state variable or parameter named "foo" is defined</s>

### CODE REVIEW

File: pynestml/cocos/co_co_no_attributes_on_continuous_port.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+from pynestml.symbols.predefined_functions import PredefinedFunctions
+from pynestml.utils.ast_utils import ASTUtils
+from pynestml.utils.logger import Logger, LoggingLevel
+from pynestml.utils.messages import Messages
+from pynestml.visitors.ast_visitor import ASTVisitor
+
+
+class CoCoNoAttributesOnContinuousPort(CoCo):
+    """
+    This context condition checker ensures that no attributes are defined on continuous time output ports.
```
Comment (pnbabu) [inline pynestml/cocos/co_co_no_attributes_on_continuous_port.py:36]: If true for all code-generation targets, could we remove the attributes from the grammar for continuous ports?

File: tests/test_cocos.py [hunk 2]
```diff
@@ -324,6 +324,29 @@ def test_valid_output_port_defined_if_emit_call(self):
...
+        assert len(Logger.get_all_messages_of_level_and_or_node(model, LoggingLevel.ERROR)) > 0
+
+    def test_valid_output_port_type_if_emit_call(self):
+        """test that an error is raised when the emit_spike() function is called with different parameter types than are defined in the spiking output port"""
+        model = self._parse_and_validate_model(os.path.join(os.path.realpath(os.path.join(os.path.dirname(__file__), 'invalid')), 'CoCoOutputPortTypeIfEmitCall-3.nestml'))
+        assert model is not None
+        assert len(Logger.get_all_messages_of_level_and_or_node(model, LoggingLevel.ERROR)) == 0
+        assert len(Logger.get_all_messages_of_level_and_or_node(model, LoggingLevel.WARNING)) > 0
+
+    def test_valid_output_port_type_continuous(self):
```
Comment (pnbabu) [inline tests/test_cocos.py:344]:
```suggestion
    def test_invalid_output_port_type_continuous(self):
```

**Review by pnbabu — CHANGES_REQUESTED** [review]

File: pynestml/cocos/co_co_no_attributes_on_continuous_port.py (hunk 1 above)
Comment (clinssen) [reply pynestml/cocos/co_co_no_attributes_on_continuous_port.py:36]: Done!

File: doc/nestml_language/nestml_language_concepts.rst [hunk 3]
```diff
@@ -907,6 +907,28 @@ Each model can only send a single type of event. The type of the event has to be
...
+
+.. code-block:: nestml
+
+   parameters:
+       weight real = 10.
+
+   update:
+       emit_spike(w, 1 ms)
+
+If spike event attributes are used, their names and types must be given as part the output port specification. The names are only used externally, for other models making reference to this one (it is allowed to have a state variable called ``weight`` and an output port attribute by the same name).
```
[the remark this answers is not in the document]
Comment (clinssen) [reply doc/nestml_language/nestml_language_concepts.rst]: Clarified, please check!

**Pull request merged**